### PR TITLE
Plugin: Remove compat layers for WP 6.4 and 6.5

### LINKED ISSUE #63115: Button Block: Remove backwards compatibility code for WP6.4
In #63082, backwards compatible code for WP6.4 was added to the Button block. This PR will be part of [Gutenberg version 18.7.1](https://github.com/WordPress/gutenberg/releases/tag/v18.7.1).

This code should be reverted _after_ the Gutenberg plugin drops support for **WP6.4** in the near future.

https://github.com/WordPress/gutenberg/blob/3687874b7975b35a94746968b6e47c2bb8ea116d/packages/block-library/src/button/index.php#L20-L33

To make sure I don't forget to do this, I'll submit an issue.

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
PR removes compat layers for WP 6.4 and 6.5. This is usually done after each release.

## How?
* Remove compat files.
* Fix any errors that appear in logs.

## Testing Instructions
* All CI checks are green.
* Smoke test editors and front-end view. They should work as before.
* There are no PHP warnings/errors in logs.

### Testing Instructions for Keyboard
Same.

### CODE REVIEW

**Comment by Mamaduka** [comment]
cc @WordPress/gutenberg-core

File: lib/load.php [hunk 1]
```diff
@@ -188,28 +122,6 @@ function gutenberg_is_experiment_enabled( $name ) {
 // Fonts API / Font Face.
 remove_action( 'plugins_loaded', '_wp_theme_json_webfonts_handler' ); // Turns off WordPress 6.0's stopgap handler.
```
Comment (Mamaduka) [inline lib/load.php]: @hellofromtonya, do we still need this?

Comment (hellofromtonya) [reply lib/load.php]: @Mamaduka no longer needed as of 6.4.0. 

[That callback was deprecated ](https://core.trac.wordpress.org/changeset/56500#file4)and its associated `add_action()` [hook](https://core.trac.wordpress.org/changeset/56500#file3) was removed in 6.4.0 via [r56500](https://core.trac.wordpress.org/changeset/56500).

Comment (Mamaduka) [reply lib/load.php]: Thank you, @hellofromtonya!

**Comment by Mamaduka** [comment]
The remaining PHPUnit test failure comes from a test introduced in #60354. Based on the description, it seems like a back-sync from the core (WP 6.5), but I can find a similar test at https://github.com/WordPress/wordpress-develop.

cc @creativecoder

File: lib/compat/wordpress-6.5/fonts/fonts.php [hunk 2]
```diff
@@ -1,485 +0,0 @@
...
-
-	foreach ( $font_files as $font_file ) {
-		$font_path = $font_dir . '/' . $font_file;
-
-		if ( file_exists( $font_path ) ) {
-			wp_delete_file( $font_path );
-		}
-	}
-}
-add_action( 'before_delete_post', 'gutenberg_before_delete_font_face', 10, 2 );
```
Comment (creativecoder) [inline lib/compat/wordpress-6.5/fonts/fonts.php:405]: As I [described here](https://github.com/WordPress/gutenberg/pull/57688#issuecomment-2259037546), I think we should leave this code in, for now. Perhaps `lib/compat/plugin/fonts.php` would be a good place for it?

Comment (Mamaduka) [reply lib/compat/wordpress-6.5/fonts/fonts.php:405]: Makes sense. Thank you, @creativecoder!

**Comment by dmsnell** [comment]
Much of the 6.6 HTML API code relies on the 6.5 compat layer. They will need to be updated and brought over as the `_6_6` variants otherwise code will crash when those `_6_5` classes are missing.

**Review by ramonjd — COMMENTED** [review]
Thanks @Mamaduka!!

I've tested the features I'm most familiar with:

✅  revision REST API related functionality (lib/compat/wordpress-6.4/class-gutenberg-rest-global-styles-revisions-controller-6-4.php)
✅  date translations (lib/compat/wordpress-6.5/script-loader.php)
✅  background CSS validation (lib/compat/wordpress-6.5/kses.php)
✅  deleting draft patterns (lib/compat/wordpress-6.4/blocks.php)

**Comment by t-hamano** [comment]
Is this PR intended to drop support for WP 6.4 and WP 6.5? If so, I think the "Requires at least" field also needs to be updated to 6.6.

Otherwise, we will likely encounter critical errors in older WordPress versions.

Here is an example of the error:

```
Fatal error: Uncaught Error: Call to undefined function get_all_registered_block_bindings_sources() in /var/www/html/wp-content/plugins/gutenberg/lib/compat/wordpress-6.6/block-bindings/pattern-overrides.php:49
Stack trace:
#0 /var/www/html/wp-includes/class-wp-hook.php(324): gutenberg_register_block_bindings_pattern_overrides_source('')
#1 /var/www/html/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#2 /var/www/html/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#3 /var/www/html/wp-settings.php(643): do_action('init')
#4 /var/www/html/wp-config.php(142): require_once('/var/www/html/w...')
#5 /var/www/html/wp-load.php(50): require_once('/var/www/html/w...')
#6 /var/www/html/wp-blog-header.php(13): require_once('/var/www/html/w...')
#7 /var/www/html/index.php(17): require('/var/www/html/w...')
#8 {main} thrown in /var/www/html/wp-content/plugins/gutenberg/lib/compat/wordpress-6.6/block-bindings/pattern-overrides.php on line 49
```


Also, I remember that the Gutenberg plugin customarily supported one version older.

**Comment by ramonjd** [comment]
> I think the "Requires at least" field also needs to be updated to 6.6.

Good point! I think @Mamaduka also raised previously that performance tests hashes might need an update?

https://github.com/WordPress/gutenberg/pull/56464/files#r1402992768

**Comment by talldan** [comment]
@Mamaduka <s>Is it right that the 6.5 folder is being removed? My understanding is that Gutenberg usually supports the last two WordPress versions (current stable and previous), so that would usually mean only removing the 6.4 folder. I can't find any written confirmation of that, though! Maybe this changed at some point, and I'm not aware.

Here's a link to one of the previous times this was done - https://github.com/WordPress/gutenberg/pull/60780</s>

Edit - actually I think you're right, here's a couple more PRs:
- https://github.com/WordPress/gutenberg/pull/53451
- https://github.com/WordPress/gutenberg/pull/56464

They both removed the folder for the version previous to the current release, as this PR does.

**Comment by talldan** [comment]
There's also some code in the button block that can be removed so I've pushed a commit - https://github.com/WordPress/gutenberg/issues/63115

**Comment by t-hamano** [comment]
Can we also remove this fallback code for WP6.4?

https://github.com/WordPress/gutenberg/blob/86735cacf73a9a52f343e457406bb717dac6213f/lib/block-supports/layout.php#L1030-L1043

**Comment by Mamaduka** [comment]
The "Requires at least" field needs to be updated to 6.5. The plugin supports the latest two major releases.

Updating the "Requires at least" field and performance hash were usually done separately.

**Comment by Mamaduka** [comment]
Here's the PR for bumping versions and perf job SHA - #64126.

**Review by talldan — APPROVED** [review]
Thanks for handling this.

I've smoke tested 6.5 and 6.6 with the build from this PR and haven't spotted any issues.

Also not seeing anything that stands out in the code, it all looks good to me. Nice job!

**Pull request merged**